Add OpenAssetIO v1.0.0-alpha.7

feltech opened the pull request:
Specify library name and version:  **OpenAssetIO/1.0.0-alpha.7**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

OpenAssetIO is a C++ library with C and Python bindings. It has a plugin system allowing so-called "manager" plugins to be loaded into a "host" application.  Hosts and plugins can be written in C++, C or Python, with OpenAssetIO marshalling between languages as necessary.

The goal is to provide a common interface for host applications to communicate with a central asset management system - a common pattern in the VFX industry.

It is an ASWF project, much like OpenColorIO, OpenEXR, OpenVDB, ..., which are also represented in ConanCenter. I am one of the maintainers of OpenAssetIO, funded by The Foundry.

OpenAssetIO is currently undergoing development and is still in the alpha stage.  We would like a ConanCenter package to give us and early adopters an easy way to begin to migrate applications/tools and flag any pain points as early as possible.

The Conan recipe has a few uncommon features:

The `fPIC` option is not supported.  Since one of the build artifacts is a Python extension module, which must be a shared library and must link to the core C++ library, we must always enable PIC for the core library.

The `cpython` package is used as a dependency. It's surprising how few (none, apparently) recipes make use of the `cpython` Conan package. The `cpython` package is not currently Conan v2 friendly, so we must use, for example, the deprecated `deps_user_info`.

The C bindings, if built as a static library, must be told to link to the C++ standard library. The way Conan's CMake generators work (`IMPORTED INTERFACE` rather than `IMPORTED STATIC`) means that this isn't done automatically. To work around this we take inspiration from  conan-io#729.

The `test_package` uses an external test repository that exercises basic linkage and imports. This is also used upstream as one of the CI checks.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/README.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.


# Review notes

## Source template

The conanfile was customised from a template in https://github.com/conan-io/conan-center-index/tree/master/docs/package_templates/cmake_package

## Testing

```
cd recipes/openassetio/all
conan create . openassetio/1.0.0-alpha.7@
rm -rf test_package/build
```
(note the `rm` line at the end - this is just to ensure no CMake cache shenanigans if building again with different options)

You can test passing options by adding e.g. `-o openassetio:shared=True -o openassetio:without_python=True`

foundrytom commented on recipes/openassetio/config.yml:
```diff
@@ -0,0 +1,3 @@
+versions:
+  "1.0.0-alpha.6":
```
[fix] Should this be:
```suggestion
  "1.0.0-alpha.7":
```

feltech replied:
Oh dear, that's worrying, definitely updated that... checking.

feltech replied:
Nope genuinely missed that one, good spot, fixed in f92d01e8

feltech commented:
Added cc0580a9 after enabling the `conan-center` hook (see checklist in PR description) and running a `conan create` to get the linter output.

Quite clever in that it pointed out one of our files (`PythonPluginSystemManagerInterfaceFactory`) has a path that is so long it'll trip up Windows.

A bit odd that it insists on having a separate `test_package` for conan v1 and v2, when the hook also says our `test_package` is conan v2 compliant, and I'm testing with conan v1, so in theory it works with either.  So I just duplicated it.

I tried this recipe with conan v2 (`conan==2.0.0b5`) and it fails on the `cpython` dependency because that recipe has not yet been updated to be conan v2 compliant.

foundrytom reviewed (approved):
looks good from my perspective - once we have the alpha 7 release in, and the URL updated in `conandata.yml`. 👍

elliotcmorris commented on recipes/openassetio/all/conanfile.py:
```diff
@@ -0,0 +1,172 @@
+        "shared": [True, False],
+        "fPIC": [True, False],
+        "without_python": [True, False],
+        "python_version": ["3.7.12", "3.8.12", "3.9.7", "3.10.0"]
```
[Question] I'd expected to see 3.7, 3.9, 3.10 (and maybe 3.11) here. Is this list something other than our supported python set?

feltech replied:
This is the list of supported versions in the `cpython` recipe on ConanCenter.

elliotcmorris reviewed (approved):
Nice one.

Just one sanity check comment, feel free to resolve it yourself and merge when ready.

feltech commented on recipes/openassetio/all/conandata.yml:
```diff
@@ -0,0 +1,5 @@
+sources:
+  "1.0.0-alpha.7":
+    # TODO(DF): Update to main org when we have a release.
```
[fix] Sort this out after release

feltech replied:
Done in [8483a67](https://github.com/OpenAssetIO/conan-center-index/pull/1/commits/8483a67f9836ba67d1313e98f151871d067189f0)

feltech commented:
I've force-pushed with an updated commit message, and integrated the message into the PR description.  Looking at the git log of the upstream repo, commit messages are very minimal so it's likely the original commit message won't survive.

Also:
- Removed the `fPIC` option in 26e63570 - as per answer to my question in https://github.com/conan-io/conan-center-index/issues/14472
- Fixed missing `homepage` attribute in d29ecc0b
- Removed an unused import in 63c53b70
